store/tikv/2pc_exec.go

```diff
@@ -0,0 +1,550 @@
+				c.handleError(err)
+				return
+			}
+		}
+
+		collector := c.getCollector(start, end)
+		for {
+			batch, err := collector.Collect()
+			if err != nil {
+				c.handleError(err)
+				return
+			}
+
+			if batch == nil {
+				return
+			}
+
+			bo, cancel := retryJob.bo.Fork()
+			job := execJob{batch, bo, cancel}
+
+			c.Add(1)
+			select {
+			case c.jobCh <- job:
+				continue
+			default:
+				c.Done()
+			}
+
+			if c.terminated() {
+				cancel()
+				return
+			}
+
+			retry, err := c.action.handleSingleBatch(c.committer, job.bo, job.batch)
+			cancel()
+			if err != nil && !retry {
+				c.handleError(err)
+				return
+			}
+			if retry {
+				start = job.batch.mutations.keys[0]
+				retryErr = err
+				break
+			}
+		}
+	}
+}
+
+func (c *execController) tryAddWorker() {
+	limit := int32(c.workerLimit)
+	for {
+		cnt := atomic.LoadInt32(&c.workerCnt)
+		if cnt >= limit {
+			return
+		}
+		if atomic.CompareAndSwapInt32(&c.workerCnt, cnt, cnt+1) {
+			go c.workerLoop()
+			return
+		}
+	}
+}
+
+func (c *execController) handleError(err error) {
+	logutil.BgLogger().Debug("2PC doActionOnBatch failed",
+		zap.Uint64("conn", c.committer.connID),
+		zap.Stringer("action type", c.action),
+		zap.Error(err),
+		zap.Uint64("txnStartTS", c.committer.startTS))
+
+	if c.cancel != nil {
+		old := atomic.SwapInt32(&c.stop, 1)
+		if old == 0 {
+			logutil.BgLogger().Debug("2PC doActionOnBatch to cancel other actions",
+				zap.Uint64("conn", c.committer.connID),
+				zap.Stringer("action type", c.action),
+				zap.Uint64("txnStartTS", c.committer.startTS))
+			c.cancel()
+		}
+	}
+
+	select {
+	case c.errCh <- err:
+	default:
+	}
+}
+
+func (c *execController) terminated() bool {
+	return atomic.LoadInt32(&c.stop) == 1
+}
+
+func (c *execController) updateSingleBatchDetail(bo *Backoffer, beforeSleep int) {
+	detail := c.committer.getDetail()
+	if detail == nil {
+		// lock operations of pessimistic-txn will let commitDetail be nil
+		return
+	}
+	delta := bo.totalSleep - beforeSleep
+	if delta <= 0 {
+		return
+	}
+	atomic.AddInt64(&detail.CommitBackoffTime, int64(delta)*int64(time.Millisecond))
+	detail.Mu.Lock()
+	detail.Mu.BackoffTypes = append(detail.Mu.BackoffTypes, bo.types...)
+	detail.Mu.Unlock()
+}
+
+type mutation struct {
+	key               []byte
+	value             []byte
+	isPessimisticLock bool
+	op                pb.Op
+}
+
+type mutations interface {
+	Iter(start, end []byte) mutationsIter
+	Len() int
+}
+
+type mutationsIter interface {
+	Next() mutation
+	WithFilter(func([]byte) bool)
+}
+
+type mutationWithRegion struct {
+	mutation
+	region RegionVerID
+}
+
+type mapWithRegionIter struct {
+	src mutationsIter
+
+	rc  *RegionCache
+	loc *KeyLocation
+	bo  *Backoffer
+}
+
+func (c *twoPhaseCommitter) mapWithRegion(bo *Backoffer, src mutationsIter) *mapWithRegionIter {
+	return &mapWithRegionIter{
+		src: src,
+		rc:  c.store.regionCache,
+		bo:  bo,
+	}
+}
+
+func (it *mapWithRegionIter) Next() (mutationWithRegion, error) {
```

Comment:
The type is `mapWithRegionIter` but it returns `mutationWithRegion`.
Why not name it `mutationWithRegionIter`?

Reply:
`map` is a verb, means this is a lazy transformation. `mapWithRegionIter{raw}` <-> `map(raw, withRegion)`.

But there's nothing wrong with changing the name, I'll update it later.